### PR TITLE
Add Unsupervised ImageFolder

### PR DESCRIPTION
**Pull request by theonekeyg**
Right now (correct me if i wrong) i can't think of better way of using ImageFolder with unlabeled datasets (e.g. "root/*.png") other than moving directory with samples into another folder and just ignore the fictional labels. So i thought it would be good to have a class that represents such types of datasets.

### CODE REVIEW

**Comment by pmeier** [comment]
Indeed you are correct, if you want a dataset of a unlabeled images, you need this crude workaround. But then I think its fair to ask: what would be the use case of such a dataset without labels?

**Comment by theonekeyg** [comment]
I can see it being used for unsupervised GAN training for image generation, unpaired image-to-image translation with 2 dataloaders representing each class. I personally needed this functionality to test GAN on celeba dataset (i'm just getting started with generative learning).

**Comment by pmeier** [comment]
Yeah, I didn't think of that. Good point!

---

I recently had a similar problem and solved it with a decorator:

```python
def unsupervise(cls):
    def onlyfirst(__getitem__):
        def __getitem__wrapper(self, idx):
            return __getitem__(self, idx)[0]

        return __getitem__wrapper

    cls.__getitem__ = onlyfirst(cls.__getitem__)

    def __init__wrapper(*args, **kwargs):
        return cls(*args, **kwargs)

    return __init__wrapper
```

You can use it like this:

```python
from torchvision.datasets import CelebA


@unsupervise
class UnsupervisedCelebA(CelebA):
    pass
```

While this does not help you if you just want all images of a folder, it allows you to _unsupervise_ an arbitrary dataset without introducing new dataset code.

**Comment by theonekeyg** [comment]
That's a really neat idea, thanks for sharing that :). It is great solution from the user perspective, but wouldn't that make more sense for library to have support at least for this common kind of datasets (moreover it is not even only specific for images)?
> 
Maybe not in the form of introducing 2 whole new classes for it, we could have a keyword in `ImageFolder` marking which type is your dataset e.g.
```python
def ImageFolder(*args, samples_only=False, **kwargs):
    if samples_only:
        return UnsupervisedDatasetFolder(*args, **kwargs)
    return DatasetFolder(*args, **kwargs)
```
The problem with this is that some other class could derive from `ImageFolder` in someone else's code, so if that solution would suffice mby add a future warning not to derive from it, or do some workaround, or something similar to this. I'd like to hear what others think of it.

**Comment by pmeier** [comment]
> It's a just bit unclear to me what is exactly wrong with the initial pull? Is it because it wasn't reviewed yet, or maintainers dislike the idea of introducing 2 new classes for it, or they don't think this functionality is necessary?

After my initial skepticism, I now think the idea behind this PR is valid and should be included. But since I have no say in this we have to wait for @fmassa s feedback if we want this and if yes how we want it.

**Comment by fmassa** [comment]
Hi,

This has been discussed in the past, see https://github.com/pytorch/vision/issues/660

I think there is a very simple solution to this (as pointed in the issue), and I'm not sure we would want to have a dataset for it.
An argument against the current implementation is that it supposes a flat hierarchy of files -- what if we have nested folders with images inside? Should we handle this case as well?

Adding a flag to `ImageFolder` to conditionally return labels or not would not be ideal IMO, as I would want to avoid having the return type of the datasets to be dependent on the constructor.

Let me know if you disagree.

**Comment by theonekeyg** [comment]
Hi, thanks for detailed reply. I guess having to write small loader every time from scratch (or making a link to folder with images) isn't that huge of a problem. 

The only argument for this PR i have is that people who isn't aware of this yet would expect pytorch to support flat hierarchy datasets (partly because of ImageFolder's uncertain name), which i think is proven by amount of related issues and questions asked. The reason they can't figure out an issue themselves i think is because they get the following misleading exception:
```python
raise (RuntimeError("Found 0 files in subfolders of: " + self.root + "\n"
                    "Supported extensions are: " + ",".join(extensions)))
```
Do you think it would be a good idea to add hint to what hierarchy should look like ("root/class/*.ext") to the exception?

**Comment by pmeier** [comment]
I think the message is quite clear. It states that it doesn't found any files in the subfolders which hints at hierarchy. If you have a look at the [documentation](https://pytorch.org/docs/stable/torchvision/datasets.html#torchvision.datasets.ImageFolder) it clearly states how the images have to be arranged. 

Thus, I think your question boils down to "can we expect the user to read the documentation before he starts using it"? IMO, yes we can.

**Comment by theonekeyg** [comment]
Nevertheless, if you search the exception, you'll find multiple issues, answers to which sometimes get highly rated (like this one https://github.com/pytorch/examples/issues/236), i think it would be nice for raised error to _straightforwardly_ hint to the most common mistake.

**Comment by theonekeyg** [comment]
Do you think its already straightforward enough?

**Comment by theonekeyg** [comment]
After thinking a little longer i think i agree with you, the problem is easily searchable anyway. I think i'll close the PR now.

**Comment by fmassa** [comment]
Thanks for the PR and the discussion @theonekeyg !

I think this is indeed a recurrent issue, and many custom solutions are possible. One proposed solution was to have a `PandasDataset`, or a `CSVDataset`, which handles an arbitrary number of fields.

Still, those solutions are only partially generic, and given the ease of implementing those, I think it might be better to give an example in how to achieve this, instead of providing a function for it.
